kernel dev guide: add component guide (#21)
tock/book: src/development/component.md

piece of kernel functionality (i.e. a capsule). You can read more technical
documentation [here](https://docs.tockos.org/kernel/component/).

In this guide we will create a component for a hypothetical system call driver
called `Notifier`. Our system call driver is going to use an alarm as a resource
and requires just one other parameter: a delay value in milliseconds. The steps
should be the same for any capsule you want to create a component for.

## Setup

This guide assumes you already have the capsule created, and ideally that you
have set it up with a board to test. Making a component then just makes it
easier to include on a new board and share among boards.

## Overview

The high-level steps required are:

1. Define the static memory required for all objects used.
2. Create a struct that holds all of the resources and configuration necessary
   for the capsules.
3. Implement `finalize()` to initialize memory and perform setup.

## Step-by-Step Guide

The steps from the overview are elaborated on here.

1. **Define the static memory required for all objects used.**

   All objects in the kernel are statically allocated, so we need to statically

>    allocate memory for the objects to live in. Now, the size of those objects
>    depends on the specific board the kernel is being compiled for. Therefore, we
>    must use a macro so that all of the types are known in advance to enable the
>    kernel to compile successfully.

Reviewer: ```suggestion
   allocate memory for the objects to live in. Due to [constraints](https://github.com/tock/tock/blob/b41ecd33a361d90820df1e290086d1b22e192e54/kernel/src/utilities/static_init.rs#L56-L80) on the macros
   Tock provides for statically allocating memory, we must contain all calls to
   allocate this memory within another macro.
```

I prefer this shorter explanation, since there are technically several reasons we need to use a macro, and cannot use generics (which would otherwise seem a suitable alternative to a macro based on the description currently provided)

   Create a file in `boards/components/src` to hold the component.

   We need to define a macro to setup our state. We will use the `static_buf!()`
   macro to help with this. In the file, create a macro with the name
   `<your capsule>_component_static`. This naming convention must be followed.

   In our hypothetical case, we need to allocate room for the notifier capsule
   and a buffer. Each capsule might need slightly different resources.

   ```rust
   #[macro_export]
   macro_rules! notifier_driver_component_static {
       ($A:ty $(,)?) => {{
           let notifier_buffer = kernel::static_buf!([u8; 16]);
           let notifier_driver = kernel::static_buf!(
               capsules_extra::notifier::NotifierDriver<'static, $A>
           );

           (notifier_buffer, notifier_driver)
       };};
   }
   ```

   Notice how the macro uses the type `$A` which is the type of the underlying
   alarm. We also use full paths to avoid errors when the macro is used. The
   macro then "returns" the two statically allocated resources.

2. **Create a struct that holds all of the resources and configuration necessary
   for the capsules.**

>    Now we create the actual component object which collects all of the resources
>    and any configuration needed to successfully setup this capsule.
> 

Reviewer 2: ```suggestion
   Now we create the actual component object which collects all of the resources
   and any configuration needed to successfully setup this capsule.

   > **Note: The component struct is _not_ long-lived.** Components are used
   > only during early-boot to create all of the objects the kernel will need. Often,
   > components need to refer to each other (e.g., our Notifier needs a reference
   > to an Alarm to set timers, and the Alarm needs a reference to our Notifier so
   > it knows where to deliver timer events). For this reason, components have a
   > two-phase approach to creating objects. First, a component struct is instantiated,
   > (i.e. `<ComponentStruct>::new()`) which can hold incomplete information.
   > Once all needed objects exist, the board can pass references that were unavailable
   > during component creation to `<ComponentStruct>::finalize()`, which will return
   > a fully instantiated peripheral object. After `finalize()` has been called, there is
   > no more use for the `ComponentStruct` and so its resources should be freed. In
   > practice, `ComponentStruct`s usually live on a stack which goes out of scope
   > before entering the kernel main loop.
```

After writing this out, I'm not actually sure this should go right here exactly, at least not all of it. The "component structs are ephemeral" idea / reminder can probably stay here; the two-phase concept explanation might be more appropriate earlier in the text? Unsure.

Reviewer: If you do keep some of the first half of this, I would edit "components need to refer to each other" to "components need to refer to the products of other components"

Author: Reading this again, this isn't true. We could just pass the static buffers to new and setup everything in one step.

   need to keep track of the delay the kernel wants to use with this capsule.

   Next we can create a constructor for this component object:

   ```rust
   impl<A: 'static + time::Alarm<'static>> NotifierDriverComponent<A> {
       pub fn new(
           board_kernel: &'static kernel::Kernel,
           driver_num: usize,
           alarm: &'static A,
           delay_ms: usize,
       ) -> AlarmDriverComponent<A> {
           AlarmDriverComponent {
               board_kernel,
               driver_num,
               alarm,
               delay_ms,
           }
       }
   }
   ```

   Note, all configuration that is required must be passed in to this `new()`
   constructor.

3. **Implement `finalize()` to initialize memory and perform setup.**

   The last step is to implement the `Component` trait and the `finalize()`
   method to actually setup the capsule.

   The general format looks like:

   ```rust
   impl<A: 'static + time::Alarm<'static>> Component for NotifierDriverComponent<A> {
       type StaticInput = (...);
       type Output = ...;

       fn finalize(self, static_buffer: Self::StaticInput) -> Self::Output {}
   }
   ```

>    We need to define that statically allocated types we need, and what this

Reviewer: ```suggestion
   We need to define what statically allocated types we need, and what this
```

   }
   ```

   Notice that the static input types must match the output of the macro. The
   output type is what we are actually creating.

   Inside the `finalize()` method we need to initialize the static memory and
   configure/setup the capsules:

   ```rust
   impl<A: 'static + time::Alarm<'static>> Component for AlarmDriverComponent<A> {
       type StaticInput = (
           &'static mut MaybeUninit<[u8; 16]>,
           &'static mut MaybeUninit<NotifierDriver<'static, $A>>,
       );
       type Output = &'static NotifierDriver<'static, A>;

       fn finalize(self, static_buffer: Self::StaticInput) -> Self::Output {
       	 let grant_cap = create_capability!(capabilities::MemoryAllocationCapability);

       	 let buf = static_buffer.0.write([0; 16]);

       	 let notifier = static_buffer.1.write(NotifierDriver::new(
       	     self.alarm,
       	     self.board_kernel.create_grant(self.driver_num, &grant_cap),
       	     buf,
       	     self.delay_ms,
       	 ));

         // Very important we set the callback client correctly.
       	 self.alarm.set_client(notifier);

       	 notifier
       }
   }
   ```

>    We initialize the memory for the static buffer, created the grant for the
>    syscall driver to use, provided the driver with the alarm resource, and
>    passing in the delay value to use. Lastly, we return a reference to the

Reviewer: ```suggestion
   We initialize the memory for the static buffer, create the grant for the
   syscall driver to use, provide the driver with the alarm resource, and
   pass in the delay value to use. Lastly, we return a reference to the
```

            board_kernel,
            driver_num,
            alarm,
            delay_ms,
        }
    }
}

impl<A: 'static + time::Alarm<'static>> Component for AlarmDriverComponent<A> {
    type StaticInput = (
        &'static mut MaybeUninit<[u8; 16]>,
        &'static mut MaybeUninit<NotifierDriver<'static, $A>>,
    );
    type Output = &'static NotifierDriver<'static, A>;

    fn finalize(self, static_buffer: Self::StaticInput) -> Self::Output {
		let grant_cap = create_capability!(capabilities::MemoryAllocationCapability);

		let buf = static_buffer.0.write([0; 16]);

		let notifier = static_buffer.1.write(NotifierDriver::new(
			self.alarm,
			self.board_kernel.create_grant(self.driver_num, &grant_cap),
			buf,
			self.delay_ms,
		));

		// Very important we set the callback client correctly.
		self.alarm.set_client(notifier);

		notifier
    }
}
```

## Usage

In a board's main.rs file to use the component:

```rust
let notifier = components::notifier::NotifierDriverComponent::new(
    board_kernel,
    capsules_core::notifier::DRIVER_NUM,
    alarm,
    100,
)
.finalize(components::notifier_driver_component_static!(nrf52840::rtc::Rtc));
```

## Wrap-Up

> Congratulations! You have create a component to easily create a resource in the

Reviewer: ```suggestion
Congratulations! You have created a component to easily create a resource in the
```
